File: python/course1_toolbox/P002_3_MaxPairwiseProductStressTest.py
import random


def get_max_pairwise_product_naive(nums):
    if len(nums) < 2:
        raise ValueError("Need at least 2 numbers")
    max_product = float("-inf")
    size = len(nums)
    for i in range(size):
        for j in range(i + 1, size):
            max_product = max(max_product, nums[i] * nums[j])
    return max_product

def get_max_pairwise_product_fast(nums):
    if len(nums) < 2:
        raise ValueError("Need at least 2 numbers")
    max_index1 = -1
    max_index2 = -1

    # find first max
    for i in range(len(nums)):
        if max_index1 == -1 or nums[max_index1] < nums[i]:
            max_index1 = i
    
    # find second max
    for i in range(len(nums)):
        if (max_index2 == -1 or nums[max_index2] < nums[i]) and max_index1 != i:
            max_index2 = i

    return nums[max_index1] * nums[max_index2]

def generate_random_array(size, max_value):
    """Generate random test case"""
    return [random.randint(0, max_value) for _ in range(size)]

def stress_test():
    print("Starting stress test...")
    print("Press ctrl+c to stop\n")

    test_count = 0

    while True:
        test_count += 1

        # Generate random test case
        # small arrays with size between 2 and 10
        size = random.randint(2, 10)
        max_value = 100_000
        nums = generate_random_array(size, max_value)

        # run both solutions on the random input
        result_naive = get_max_pairwise_product_naive(nums)
        restult_fast = get_max_pairwise_product_fast(nums)

        # Compare restults 
        if result_naive != restult_fast:
            print("MISMATCH FOUND")
            print(f"Test case: {nums}")
            print(f"naive result: {result_naive}")
            print(f"fast result: {restult_fast}")
            print(f"\nTest failed after {test_count} iterations")
            break

        # Print Progress every 1000 tests
        if test_count % 1000 == 0:
            print(f"Passed {test_count} tests...")

if __name__ == "__main__":
    try:
        stress_test()
    except KeyboardInterrupt:
        print("\n\nStress test stopped by user")

File: java/ucsd-dsa/course1_toolbox/src/main/java/edu/ucsd/dsa/toolbox/module1/P002_3_MaxPairwiseProductStressTest.java
package edu.ucsd.dsa.toolbox.module1;

import java.util.Random;

/**
 * Stress Testing: generates random test cases, runs 2 different solutions and compares the results
 *
 * Implementation details:
 * 01. the naive solution method
 * 02. the fast solution method
 * 03. a method to generate random input
 * 04. in the main method, write infinite loop to run the 2 solutions on the input and compare the results
 *
 * utility method for printing is added
 */
public class P002_3_MaxPairwiseProductStressTest {

    /**
     * Naive solution O(n^2)
     * @param nums the input array of numbers
     * @return the max product of 2 numbers in the input array
     */
    private static long getMaxPairwiseProductNaive(long[] nums) {
        if(nums == null || nums.length < 2) throw new IllegalArgumentException("Need at least 2 numbers");
        long maxProduct = Long.MIN_VALUE;
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                maxProduct = Math.max(maxProduct, nums[i] * nums[j]);
            }
        }
        return maxProduct;
    }

    private static long getMaxPairwiseProductFast(long[] nums) {
        if(nums == null || nums.length < 2) throw new IllegalArgumentException("Need at least 2 numbers");
        int maxIndex1 = -1;
        int maxIndex2 = -1;

        // find the first max
        for (int i = 0; i < nums.length; i++) {
            if (maxIndex1 == -1 || nums[i] > nums[maxIndex1]) {
                maxIndex1 = i;
            }
        }

        // find the second max
        for (int i = 0; i < nums.length; i++) {
            if ((maxIndex2 == -1 || nums[i] > nums[maxIndex2]) && i != maxIndex1) {
                maxIndex2 = i;
            }
        }

        // casting to int is added intentionally to make the test fail with integer overflow,
        // this should produce mismatch between long vs int values when the long value from the naive
        // method is way bigger than the max int value
        // remove "(int)" to make the test passes
        return (int)nums[maxIndex1] *  (int)nums[maxIndex2];
    }

    //generate random test cases
    private static long[] generateRandomArray(int size, long maxValue){
        Random random = new Random();
        long[] nums = new long[size];
        for (int i = 0; i < size; i++) {
            nums[i] = random.nextLong(maxValue + 1);
        }
        return nums;
    }

    // print array helper
    private static void printArray(long[] nums) {
        System.out.println("[");
        for (int i = 0; i < nums.length; i++) {
            System.out.print(nums[i]);
            if (i != nums.length - 1) System.out.print(", ");
        }
        System.out.println("]");
    }

    public static void main(String[] args) {
        System.out.println("Staring Stress Test...");
        System.out.println("Press Ctrl + c to stop\n");

        int testCount = 0;

        while (true){
            testCount++;

            int size = 2 + (int)(Math.random() * 9); // random number between 2 and 10
            long maxValue = 100_000;
            long[] nums = generateRandomArray(size, maxValue);

            // Run both solutions
            long resultNaive = getMaxPairwiseProductNaive(nums);
            long resultFast = getMaxPairwiseProductFast(nums);

            // Compare results
            if(resultNaive != resultFast){
                System.out.println("MISMATCH FOUND!");
                System.out.println("Test Case: ");
                printArray(nums);
                System.out.println("Result Naive: " + resultNaive);
                System.out.println("Result Fast: " + resultFast);
                System.out.println("\nTest failed after " + testCount + " iterations");
                break;
            }

            // print progress every 1000 tests
            if(testCount % 1000 == 0){
                System.out.println("Passed " + testCount + " tests...");
            }
        }
    }
}
